cpp/daal/src/externals/service_service_mkl.h

```diff
@@ -67,19 +78,71 @@ struct MklService
         return i;
     }
 
-    static int serv_strncpy_s(char * dest, size_t dmax, const char * src, size_t slen) { return fpk_serv_strncpy_s(dest, dmax, src, slen); }
+    static int serv_strncpy_s(char * dest, size_t dmax, const char * src, size_t slen)
+    {
+        if (dmax < slen) return static_cast<int>(ENOMEM);
+        strncpy(dest, src, slen);
+        return 0;
+        // TODO: safe funtion
+        // return strncpy_s(dest, dmax, src, slen);
+    }
 
-    static int serv_strncat_s(char * dest, size_t dmax, const char * src, size_t slen) { return fpk_serv_strncat_s(dest, dmax, src, slen); }
+    static int serv_strncat_s(char * dest, size_t dmax, const char * src, size_t slen)
+    {
+        if (dmax < slen) return static_cast<int>(ENOMEM);
+        strncat(dest, src, slen);
+        return 0;
+        // TODO: safe funtion
+        // return strncat_s(dest, dmax, src, slen);
+    }
 
-    static float serv_string_to_float(const char * nptr, char ** endptr) { return __FPK_string_to_float(nptr, endptr); }
+    static double serv_string_to_double(const char * nptr, char ** endptr)
+    {
+        const char * cur = nptr;
+        for (; isdigit(*cur) || *cur == '-' || *cur == 'e' || *cur == 'E' || *cur == '.'; ++cur)
+            ;
+        if (endptr) *endptr = const_cast<char *>(cur);
+        size_t size = cur - nptr;
+        // TODO replace with static buffer
+        char * buffer = static_cast<char *>(malloc(size + 1));
+        for (size_t i = 0; i < size; ++i) buffer[i] = nptr[i];
+        buffer[size] = '\0';
+        double val   = atof(buffer);
+        free(buffer);
+        return val;
+    }
```

Comment:
It is strongly encouraged to use [`std::from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars) in modern C++. However - you can still use [a much safer alternative](https://en.cppreference.com/w/cpp/string/basic_string/stof) since C++11.

Reply:
```suggestion
    static double serv_string_to_double(const char * nptr, char ** endptr)
    {
        std::size_t npos = 0ul;
        double result = std::stod({ nptr }, &npos);
        (*endptr) = static_cast<char*>(nptr + npos);
        return result;
    }
```